Introduce a class to represent a "closed set of concrete specs"

alalazo opened the pull request:
There are many places in Spack where we need to operate with a set of consistent spec objects, i.e. a set of specs where:
1. A single DAG hash refers to a single object in memory
2. Each spec has no references to other specs outside the set

This is, for instance, the case when we need to reconstruct a concrete spec during concretization, in particular if that spec is provided by more than one source of "reusable" specs (see #39590).

Another case is when we concretize or reconstruct environments. In those cases we have custom ad-hoc code to ensure consistency in memory. For concretization that would be:

https://github.com/spack/spack/blob/2f24aeb7f690a7cc73996479b0028a273e91d9df/lib/spack/spack/environment/environment.py#L1601-L1624

while when reading an environment from a lockfile:

https://github.com/spack/spack/blob/2f24aeb7f690a7cc73996479b0028a273e91d9df/lib/spack/spack/environment/environment.py#L2239-L2257

This PR is an attempt at collecting these similar operations together into a class, which can be reused in different places within the codebase. This class is a _mapping from hashes to concrete specs_ that:
1. Ensures consistency of the specs being managed
2. Provides methods to add, delete, query specs

The idea is to refine an API for the class in this PR, and use it for `Environments`, and apply it in later PRs also to Databases and Indexes.

alalazo committed:
Use ConcreteSpecsByHash to unify environments

haampie commented on lib/spack/spack/environment/environment.py:
```diff
@@ -1527,33 +1527,14 @@ def _concretize_separately(self, tests=False):
         # Add specs in original order
         batch.sort(key=lambda x: x[0])
-        by_hash = {}  # for attaching information on test dependencies
+        unified_specs = spack.solver.asp.ConcreteSpecsByHash()
```
I don't think this is correct cause the environment already has old concrete specs? Also the term unify is overloaded. Maybe deduplicate (by dag hash)?

alalazo replied:
> I don't think this is correct cause the environment already has old concrete specs?

Fair point. Makes me wonder if we should use this class as an internal member of the environment.

alalazo committed:
Add query and delete operations to ConcreteSpecsByHash

alalazo committed:
Extract a method to initialize spec related state in environments

alalazo committed:
Introduce Environment.all_specs_by_hash

alalazo committed:
environment: rename "specs_by_hash" to "root_specs_by_hash"

alalazo committed:
environment: reuse code to build the env from lockfile

alalazo committed:
environment: fix deconcretize method

alalazo committed:
environment: remove dead code

alalazo committed:
environment: simplify all_hashes

alalazo committed:
environment: remove a traverse in get_by_hash

alalazo committed:
environment: remove a traverse in matching_spec

alalazo commented on lib/spack/spack/environment/environment.py:
```diff
@@ -1346,15 +1345,26 @@ def concretize(self, force=False, tests=False):
         if dag_hash not in self.concretized_order:
-            del self.specs_by_hash[dag_hash]
+            del self.root_specs_by_hash[dag_hash]
+            try:
+                self.all_specs_by_hash.delete(spec, transitive=False)
+            except ValueError:
+                pass
```
Note to self: this isn't completely correct

alalazo committed:
environment: remove unneeded local variable

alalazo committed:
Add docstrings for query and delete methods

alalazo committed:
Move class to its own Python module

alalazo committed:
ConcreteSpecsByHash: add a garbage_collection method

This method receives a list of concrete specs to keep, and removes
everything else.

haampie commented:
I'm not immediately a fan of the fact that this always duplicates.

1. It's redundant in many cases (and then it's just a performance penalty)
2. I don't know if we store any properties on `Spec` object like we do on `PackageBase` instances, like `installed_from_binary_cache`, that would get lost after `_dup`.

If you wanna go this way in environments, you should use `ConcreteSpecsByHash` as a spec builder, instead of first building the Spec objects graph, only to tear them apart and reconstruct them in `ConcreteSpecsByHash` again.

The overhead of deserializing an environment is really felt in CI with e.g. the e4s env, any command `spack -e <env> ...` is ultra slow already.
